Unmarshaling arrays panics

**Describe the bug**
Unmarshaling panics:
```
panic: reflect.MakeSlice of non-slice type

goroutine 1 [running]:
reflect.MakeSlice(0x566460, 0x513e20, 0x4, 0x4, 0x4ead48, 0x510560, 0x8)
        /usr/local/go/src/reflect/value.go:2256 +0x1d8
github.com/pelletier/go-toml.(*Decoder).valueFromOtherSlice(0xc00011ff10, 0x566460, 0x513e20, 0xc000022240, 0x4, 0x4, 0xc00011f838, 0xc00011f6a0, 0x4c99b3, 0xc0001101e0, ...)
        /projects/go-toml/marshal.go:786 +0x5f
github.com/pelletier/go-toml.(*Decoder).valueFromToml(0xc00011ff10, 0x566460, 0x513e20, 0x50dee0, 0xc00000c260, 0xc00000c280, 0x191, 0x0, 0x0, 0xc000014200, ...)
        /projects/go-toml/marshal.go:856 +0x527
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc00011ff10, 0x566460, 0x51c080, 0xc000068480, 0xc00011fe10, 0x199, 0xc00005c2a0, 0x0, 0xc00011fdb0, 0x4f2dba)
        /projects/go-toml/marshal.go:694 +0x1174
github.com/pelletier/go-toml.(*Decoder).unmarshal(0xc00011ff10, 0x50a900, 0xc0000141f8, 0xc000068480, 0x0)
        /projects/go-toml/marshal.go:638 +0x275
github.com/pelletier/go-toml.(*Tree).Unmarshal(...)
        /projects/go-toml/marshal.go:525
main.main()
        /projects/go-toml/cmd/example/main.go:18 +0x13f
```

**To Reproduce**
```go
type element struct {
	Color [4]uint8
}

tree, _ := toml.Load(`color = [1,2,3,4]`)

var e element
err := tree.Unmarshal(&e)
if err != nil {
	fmt.Println("unmarshal error:", err)
}
```
**Expected behavior**
There should be no panic. 

**Versions**
 - go-toml: last version of master (2b8e33f)
 - go: 1.14
 - operating system: Linux

**Additional context**
None.

--- a/marshal.go
+++ b/marshal.go
@@ -768,7 +768,11 @@ func (d *Decoder) valueFromTree(mtype reflect.Type, tval *Tree, mval1 *reflect.V
 
 // Convert toml value to marshal struct/map slice, using marshal type
 func (d *Decoder) valueFromTreeSlice(mtype reflect.Type, tval []*Tree) (reflect.Value, error) {
-	mval := reflect.MakeSlice(mtype, len(tval), len(tval))
+	mval, err := makeSliceOrArray(mtype, len(tval))
+	if err != nil {
+		return mval, err
+	}
+
 	for i := 0; i < len(tval); i++ {
 		d.visitor.push(strconv.Itoa(i))
 		val, err := d.valueFromTree(mtype.Elem(), tval[i], nil)
@@ -783,7 +787,11 @@ func (d *Decoder) valueFromTreeSlice(mtype reflect.Type, tval []*Tree) (reflect.
 
 // Convert toml value to marshal primitive slice, using marshal type
 func (d *Decoder) valueFromOtherSlice(mtype reflect.Type, tval []interface{}) (reflect.Value, error) {
-	mval := reflect.MakeSlice(mtype, len(tval), len(tval))
+	mval, err := makeSliceOrArray(mtype, len(tval))
+	if err != nil {
+		return mval, err
+	}
+
 	for i := 0; i < len(tval); i++ {
 		val, err := d.valueFromToml(mtype.Elem(), tval[i], nil)
 		if err != nil {
@@ -797,15 +805,34 @@ func (d *Decoder) valueFromOtherSlice(mtype reflect.Type, tval []interface{}) (r
 // Convert toml value to marshal primitive slice, using marshal type
 func (d *Decoder) valueFromOtherSliceI(mtype reflect.Type, tval interface{}) (reflect.Value, error) {
 	val := reflect.ValueOf(tval)
+	length := val.Len()
 
-	lenght := val.Len()
-	mval := reflect.MakeSlice(mtype, lenght, lenght)
-	for i := 0; i < lenght; i++ {
+	mval, err := makeSliceOrArray(mtype, length)
+	if err != nil {
+		return mval, err
+	}
+
+	for i := 0; i < length; i++ {
 		val, err := d.valueFromToml(mtype.Elem(), val.Index(i).Interface(), nil)
 		if err != nil {
 			return mval, err
 		}
 		mval.Index(i).Set(val)
+	}
+	return mval, nil
+}
+
+// Create a new slice or a new array with specified length
+func makeSliceOrArray(mtype reflect.Type, tLength int) (reflect.Value, error) {
+	var mval reflect.Value
+	switch mtype.Kind() {
+	case reflect.Slice:
+		mval = reflect.MakeSlice(mtype, tLength, tLength)
+	case reflect.Array:
+		mval = reflect.New(reflect.ArrayOf(mtype.Len(), mtype.Elem())).Elem()
+		if tLength > mtype.Len() {
+			return mval, fmt.Errorf("unmarshal: TOML array length (%v) exceeds destination array length (%v)", tLength, mtype.Len())
+		}
 	}
 	return mval, nil
 }
@@ -955,7 +982,7 @@ func (d *Decoder) valueFromToml(mtype reflect.Type, tval interface{}, mval1 *ref
 				ival := mval1.Elem()
 				return d.valueFromToml(mval1.Elem().Type(), t, &ival)
 			}
-		case reflect.Slice:
+		case reflect.Slice, reflect.Array:
 			if isOtherSequence(mtype) && isOtherSequence(reflect.TypeOf(t)) {
 				return d.valueFromOtherSliceI(mtype, t)
 			}
@@ -1025,11 +1052,7 @@ func tomlOptions(vf reflect.StructField, an annotation) tomlOpts {
 
 func isZero(val reflect.Value) bool {
 	switch val.Type().Kind() {
-	case reflect.Map:
-		fallthrough
-	case reflect.Array:
-		fallthrough
-	case reflect.Slice:
+	case reflect.Slice, reflect.Array, reflect.Map:
 		return val.Len() == 0
 	default:
 		return reflect.DeepEqual(val.Interface(), reflect.Zero(val.Type()).Interface())

--- a/marshal_test.go
+++ b/marshal_test.go
@@ -2258,7 +2258,7 @@ func TestUnmarshalPreservesUnexportedFields(t *testing.T) {
 
 	[[slice1]]
 	exported1 = "visible3"
-	
+
 	[[slice1]]
 	exported1 = "visible4"
 
@@ -3121,6 +3121,20 @@ type sliceStruct struct {
 	StructSlicePtr *[]basicMarshalTestSubStruct `  toml:"struct_slice_ptr"  `
 }
 
+type arrayStruct struct {
+	Slice          [4]string                     `  toml:"str_slice"  `
+	SlicePtr       *[4]string                    `  toml:"str_slice_ptr"  `
+	IntSlice       [4]int                        `  toml:"int_slice"  `
+	IntSlicePtr    *[4]int                       `  toml:"int_slice_ptr"  `
+	StructSlice    [4]basicMarshalTestSubStruct  `  toml:"struct_slice"  `
+	StructSlicePtr *[4]basicMarshalTestSubStruct `  toml:"struct_slice_ptr"  `
+}
+
+type arrayTooSmallStruct struct {
+	Slice       [1]string                    `  toml:"str_slice"  `
+	StructSlice [1]basicMarshalTestSubStruct `  toml:"struct_slice"  `
+}
+
 func TestUnmarshalSlice(t *testing.T) {
 	tree, _ := LoadBytes(sliceTomlDemo)
 	tree, _ = TreeFromMap(tree.ToMap())
@@ -3165,6 +3179,75 @@ func TestUnmarshalSliceFail2(t *testing.T) {
 		t.Error("expect err:(1, 1): Can't convert 1(int64) to string but got ", err)
 	}
 
+}
+
+func TestUnmarshalArray(t *testing.T) {
+	var tree *Tree
+	var err error
+
+	tree, _ = LoadBytes(sliceTomlDemo)
+	var actual1 arrayStruct
+	err = tree.Unmarshal(&actual1)
+	if err != nil {
+		t.Error("shound not err", err)
+	}
+
+	tree, _ = TreeFromMap(tree.ToMap())
+	var actual2 arrayStruct
+	err = tree.Unmarshal(&actual2)
+	if err != nil {
+		t.Error("shound not err", err)
+	}
+
+	expected := arrayStruct{
+		Slice:          [4]string{"Howdy", "Hey There"},
+		SlicePtr:       &[4]string{"Howdy", "Hey There"},
+		IntSlice:       [4]int{1, 2},
+		IntSlicePtr:    &[4]int{1, 2},
+		StructSlice:    [4]basicMarshalTestSubStruct{{"1"}, {"2"}},
+		StructSlicePtr: &[4]basicMarshalTestSubStruct{{"1"}, {"2"}},
+	}
+	if !reflect.DeepEqual(actual1, expected) {
+		t.Errorf("Bad unmarshal: expected %v, got %v", expected, actual1)
+	}
+	if !reflect.DeepEqual(actual2, expected) {
+		t.Errorf("Bad unmarshal: expected %v, got %v", expected, actual2)
+	}
+}
+
+func TestUnmarshalArrayFail(t *testing.T) {
+	tree, _ := TreeFromMap(map[string]interface{}{
+		"str_slice": []string{"Howdy", "Hey There"},
+	})
+
+	var actual arrayTooSmallStruct
+	err := tree.Unmarshal(&actual)
+	if err.Error() != "(0, 0): unmarshal: TOML array length (2) exceeds destination array length (1)" {
+		t.Error("expect err:(0, 0): unmarshal: TOML array length (2) exceeds destination array length (1) but got ", err)
+	}
+}
+
+func TestUnmarshalArrayFail2(t *testing.T) {
+	tree, _ := Load(`str_slice=["Howdy","Hey There"]`)
+
+	var actual arrayTooSmallStruct
+	err := tree.Unmarshal(&actual)
+	if err.Error() != "(1, 1): unmarshal: TOML array length (2) exceeds destination array length (1)" {
+		t.Error("expect err:(1, 1): unmarshal: TOML array length (2) exceeds destination array length (1) but got ", err)
+	}
+}
+
+func TestUnmarshalArrayFail3(t *testing.T) {
+	tree, _ := Load(`[[struct_slice]]
+String2="1"
+[[struct_slice]]
+String2="2"`)
+
+	var actual arrayTooSmallStruct
+	err := tree.Unmarshal(&actual)
+	if err.Error() != "(3, 1): unmarshal: TOML array length (2) exceeds destination array length (1)" {
+		t.Error("expect err:(3, 1): unmarshal: TOML array length (2) exceeds destination array length (1) but got ", err)
+	}
 }
 
 func TestDecoderStrict(t *testing.T) {

